lib/active_merchant/billing/gateways/shift4.rb

```diff
@@ -0,0 +1,300 @@
+          post[:transaction][:purchaseCard][:productDescriptors] = options[:purchase_card][:product_descriptors]
+        end
+        post[:overrideBusDate] = options[:override_bus_date] if options[:override_bus_date].present?
+        if options[:amex].present?
+          post[:amex] = {}
+          post[:amex][:propertyCode] = options[:amex][:property_code]
+        end
+        if options[:auto].present?
+          post[:auto] = {}
+          post[:auto][:estimatedDays] = options[:auto][:estimated_days]
+        end
+        if options[:card_on_file].present?
+          post[:cardOnFile] = {}
+          post[:cardOnFile][:indicator] = options[:card_on_file][:indicator]
+          post[:cardOnFile][:usageIndicator] = options[:card_on_file][:usage_indicator]
+          post[:cardOnFile][:scheduledIndicator] = options[:card_on_file][:scheduled_indicator]
+          post[:cardOnFile][:indicator] = options[:card_on_file][:transaction_id]
+        end
+        if options[:hotel].present?
+          post[:hotel] = {}
+          post[:hotel][:estimatedDays] = options[:hotel][:estimated_days]
+        end
+        if options[:pin].present?
+          post[:pin] = {}
+          post[:pin][:block] = options[:pin][:block]
+          post[:pin][:ksn] = options[:pin][:ksn]
+        end
+      end
+
+      def add_device(post, options)
+        if options[:device].present?
+          post[:device] = {}
+          post[:device][:manufacturer] = options[:device][:manufacturer]
+          post[:device][:model] = options[:device][:model]
+          post[:device][:promptPostalCode] = options[:device][:prompt_postal_code]
+          post[:device][:promptSecurityCode] = options[:device][:prompt_security_code]
+          post[:device][:promptStreetNumber] = options[:device][:prompt_street_number]
+          post[:device][:terminalId] = options[:device][:terminal_id]
+        end
+      end
+
+      def add_emv(post, options)
+        if options[:emv].present?
+          post[:emv] = {}
+          post[:emv][:tlvData] = options[:emv][:tlv_data]
+        end
+      end
+
+      def add_p2pe(post, options)
+        if options[:p2pe].present?
+          post[:p2pe] = {}
+          post[:p2pe][:format] = options[:p2pe][:format]
+          post[:p2pe][:ksn] = options[:p2pe][:ksn]
+        end
+      end
+
+      def add_card(post, credit_card, options)
+        post[:card] = {}
+        if credit_card.respond_to?(:number)
+          post[:card][:expirationDate] = "0#{credit_card.month}#{credit_card.year.to_s[2..4]}"
+          post[:card][:number] = credit_card.number
+          post[:card][:entryMode] = options[:entry_mode]
+          post[:card][:present] = options[:present]
+          post[:card][:securityCode] = {}
+        else
+          post[:card] = {} if post[:card].nil?
+          post[:card][:token] = {}
+          post[:card][:token][:value] = credit_card
+        end
+        if options[:card].present?
```

Comment:
I don't think we should expect to pass any card information via the `options` hash.

Reply:
@naashton Card object has other info too that is
          post[:card][:entryMode] = options[:card][:entry_mode]
          post[:card][:present] = options[:card][:present]
          post[:card][:securityCode] = {} unless post[:card][:securityCode]
          post[:card][:securityCode][:indicator] = options[:card][:indicator] if options[:card][:indicator]
          post[:card][:securityCode][:value] = options[:card][:value] if options[:card][:value]
          post[:card][:securityCode] = {}

Reply:
I think it's fine to expect `entry_mode`, `present`, and maybe `indicator` (not quite sure what this is) to be passed in the `options` hash, but for security reasons, we should grab the `security_code` and `value` (I'm assuming this correlates to `credit_card.number`) from our `CreditCard` object.